ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -209,6 +209,13 @@ public int compare(IndexEntry e1, IndexEntry e2) {
       log.setActiveSegment(getCurrentEndOffset().getName());
       logEndOffsetOnStartup = log.getEndOffset();
 
+      // Validate store.index.max.number.of.inmem.elements to prevent incorrect Journal size
+      if (!getIndexSegments().isEmpty() &&
+          config.storeIndexMaxNumberOfInmemElements < getIndexSegments().lastEntry().getValue().getNumberOfItems()) {
+        throw new StoreException("store.index.max.number.of.inmem.elements is less than the number of entries in the last index segment",
```

Comment:
also print the values (config value and the size of the index segment)

Reply:
Also, should we compare the config or the actual size of the journal? I think it should be the latter

Reply:
On a second thought I think so too. Otherwise it'll be difficult to lower the `store.index.max.number.of.inmem.elements` config. You will have to be betting on the fact that the last index segment is unpopulated.

Reply:
That's right - but considering that `PersistentIndex` has a constructor that takes in the `Journal`, it is possible (as far as the API goes), to lower the config but keep it correct by passing in an oversized journal